### PR TITLE
PERF: fast non-unique indexing

### LINKED ISSUE #15364: Huge performance hit in get_indexer_non_unique
Let `df0` and `df1` be indexed data frames. Assume that at least one of them has a non-unique index. If I'm not mistaken, the function [#L309](https://github.com/pandas-dev/pandas/blob/3853fe6d2b884f186b93933ef53ff5e475c2d80c/pandas/index.pyx#L309) will always be called upon  the operation:

```python
df0.loc[df1.index]
```

Now imagine that both of indexes are reasonably large (lets say at least one millions of records). The operations at lines

- [#L341](https://github.com/pandas-dev/pandas/blob/3853fe6d2b884f186b93933ef53ff5e475c2d80c/pandas/index.pyx#L341)
- [#L342](https://github.com/pandas-dev/pandas/blob/3853fe6d2b884f186b93933ef53ff5e475c2d80c/pandas/index.pyx#L342)
- [#L351](https://github.com/pandas-dev/pandas/blob/3853fe6d2b884f186b93933ef53ff5e475c2d80c/pandas/index.pyx#L351)

will be of order `O(n^2*log(n)^2)` even when both indices are sorted.

Am I missing something? Let me know if the above reasoning is right so I can help with it. If both indexes are sorted, It looks like that function can be as fast as `O(n)` simply by looping over the `values` and `targets` arrays simultaneously. (Assuming that the index duplication is very small compared to `n`.

### PR DESCRIPTION
**Pull request by horta**
closes #15364 

This is essentially the same pull request as https://github.com/pandas-dev/pandas/pull/15372 but now from a proper branch.

### CODE REVIEW

File: pandas/indexes/base.py [hunk 1]
```diff
@@ -2530,10 +2530,40 @@ def get_indexer_non_unique(self, target):
         if self.is_all_dates:
             self = Index(self.asi8)
             tgt_values = target.asi8
+            src_values = self.asi8
         else:
             tgt_values = target._values
+            src_values = self._values
+
```
Comment (jreback) [inline pandas/indexes/base.py]: this is quite messy, what are you trying to do here?

File: pandas/index.pyx [hunk 2]
```diff
@@ -371,6 +482,16 @@ cdef class IndexEngine:
 
         return result[0:count], missing[0:count_missing]
 
+    def get_indexer_non_unique_orderable(self, ndarray targets,
+                                         int64_t[:] idx0,
+                                         int64_t[:] idx1):
```
Comment (jreback) [inline pandas/index.pyx]: this is a massive amount of added code, what exactly are you doing?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/indexes/base.py (hunk 1 above)
Comment (Horta) [inline pandas/indexes/base.py]: Agreed! Just factorised that.

Comment (jreback) [inline pandas/indexes/base.py]: you are adding much more code. Where exactly does this routine fail? The only scenario I could see is with actual mixed types (which I am happy to raise on when you have a non-unique index). 

Further it looks like you are duplicating lots of functionailty that already exists, see pandas/core/algorithms.py w.r.t. counting / sorting / factorizing.

Comment (Horta) [inline pandas/indexes/base.py]: I didn't find routines un pandas/core/algorithms.py that I could use. I have now described the functions in index.pyx so I hope it is now clearer.

Yes, I was having problem with mixed types. For example,

```python
v = np.array([1, 'danilo'], object)
v[0] < v[1]
```
raises a `TypeError` exception.

File: pandas/index.pyx (hunk 2 above)
Comment (Horta) [inline pandas/index.pyx]: Notice that I have removed the code regarding lists. Most of the code consists in two functions: `_count` and `_map`. They implement two very specific algorithms that I couldn't find in the above-mentioned file.

**Comment by Horta** [comment]
Benchmark:

```python
import numpy as np
import pandas as pd
from time import time
from pandas import Index

n = 10000
r = 100


for j in range(1, 11):
    i = Index(list(range(j*n)) * r)
    slic = i[0:j*n]
    start = time()
    i.get_indexer_non_unique(slic)
    print(time() - start)
```

Elapsed time in seconds:

```
Previous | New
---------------
    0.92 | 0.18
    3.03 | 0.37
    6.25 | 0.64
   10.90 | 0.79
   17.82 | 1.00
   25.07 | 1.16
   31.49 | 1.33
   40.97 | 1.53
   51.54 | 2.03
   60.64 | 2.18
```

File: pandas/index.pyx [hunk 3]
```diff
@@ -44,6 +44,190 @@ PyDateTime_IMPORT
...
     int PySlice_Check(object)
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.initializedcheck(False)
+cdef _count(ndarray values, ndarray targets, int64_t[:] idx0, int64_t[:] idx1,
+            int64_t[:] mapping_count, int64_t[:] missing_count):
+    """
+    Compute the number of times a `targets` value is in `values`.
+
```
Comment (jreback) [inline pandas/index.pyx]: this is exactly value_counts

**Comment by jreback** [comment]
@Horta you are adding more and more code. I want LESS code. you are rewriting the world here.

File: pandas/index.pyx [hunk 4]
```diff
@@ -44,6 +44,190 @@ PyDateTime_IMPORT
...
+        mapping_count[idx1[j]] += 1
+        missing_count[idx1[j]] = 1
+        j += 1
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.initializedcheck(False)
+cdef _map(ndarray values, ndarray targets, int64_t[:] idx0, int64_t[:] idx1,
+            int64_t[:] start_mapping, int64_t[:] start_missing,
+            int64_t[:] mapping, int64_t[:] missing):
```
Comment (jreback) [inline pandas/index.pyx]: this signature is WAY too complicated.

File: pandas/index.pyx [hunk 5]
```diff
@@ -44,6 +44,190 @@ PyDateTime_IMPORT
...
+    while j < n_t:
+
+        mapping[start_mapping[idx1[j]]] = -1
+        missing[start_missing[idx1[j]]] = idx1[j]
+        j += 1
+
+def _map_targets_to_values(values, targets, idx0, idx1):
+    """
+    Map `targets` values to `values` positions.
+
```
Comment (jreback) [inline pandas/index.pyx]: this looks like a take, but really not sure what you are doing

File: pandas/index.pyx [hunk 6]
```diff
@@ -371,6 +555,16 @@ cdef class IndexEngine:
 
         return result[0:count], missing[0:count_missing]
 
+    def get_indexer_non_unique_orderable(self, ndarray targets,
+                                         int64_t[:] idx0,
+                                         int64_t[:] idx1):
+
+        cdef:
+            ndarray values
+
```
Comment (jreback) [inline pandas/index.pyx]: you can add 1 function in cython, anything else is WAY too complicated.

File: pandas/indexes/base.py [hunk 7]
```diff
@@ -2530,10 +2530,23 @@ def get_indexer_non_unique(self, target):
...
+        is_mono0 = self.is_monotonic_increasing
+        is_mono1 = target.is_monotonic_increasing
+        (orderable, idx0, idx1) = _order_them(is_mono0, src_values, is_mono1,
```
Comment (jreback) [inline pandas/indexes/base.py]: if you insist on ordering, then simply order an unordered index, get the results and take the original. this is so much complexity.

File: pandas/indexes/base.py [hunk 8]
```diff
@@ -3875,3 +3888,43 @@ def _trim_front(strings):
...
+
+def _order_it(is_mono, x):
+    """Tries to sort a sequence."""
+    if is_mono:
+        return (True, np.arange(len(x), dtype=np.int64))
+    try:
+        indices = np.argsort(x, kind='mergesort')
+    except TypeError:
+        return (False, None)
+    return (True, indices)
```
Comment (jreback) [inline pandas/indexes/base.py]: do ALL of this in the cython function. you are needlessly splitting things up.

**Comment by jreback** [comment]
pls add an asv.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by Horta** [comment]
@jreback You are such an ass. Closing here as I'm not helping it anymore.

File: pandas/indexes/base.py (hunk 7 above)
Comment (jreback) [inline pandas/indexes/base.py]: @Horta we don't tolerate that type of language

**Comment by jreback** [comment]
@Horta we don't tolerate that type of language

**Pull request closed without merging**